phd-tests/tests/src/boot_order.rs

```diff
@@ -0,0 +1,533 @@
+        .await?
+        .is_empty()
+    {
+        warn!(
+            "guest environment already has a BootFFFF entry; \
+            is this not a fresh image?"
+        );
+    }
+
+    let boot_num: u16 = {
+        let bytes = read_efivar(&vm, BOOT_CURRENT_VAR).await?;
+        u16::from_le_bytes(bytes.try_into().unwrap())
+    };
+
+    // The entry we booted from is clearly valid, so we should be able to insert
+    // a few duplicate entries. We won't boot into them, but if something bogus
+    // happens and we did boot one of these, at least it'll work and we can
+    // detect the misbehavior.
+    //
+    // But here's a weird one: if we just append these to the end, on reboot
+    // they'll be moved somewhat up the boot order. This occurrs both if setting
+    // variables through `efibootmgr` or by writing to
+    // /sys/firmware/efi/efivars/BootOrder-* directly. As an example, say we had
+    // a boot order of "0004,0001,0003,0000" where boot options were as follows:
+    // * 0000: UiApp
+    // * 0001: PCI device 4, function 0
+    // * 0003: EFI shell (Firmware volume+file)
+    // * 0004: Ubuntu (HD partition 15, GPT formatted)
+    //
+    // If we duplicate entry 4 to new options FFF0 and FFFF, reset the boot
+    // order to "0004,0001,0003,0000,FFF0,FFFF", then reboot the VM, the boot
+    // order when it comes back up will be "0004,0001,FFF0,FFFF,0003,0000".
+    //
+    // This almost makes sense, but with other devices in the mix I've seen
+    // reorderings like `0004,0001,<PCI 16.0>,0003,0000,FFF0,FFFF` turning into
+    // `0004,0001,FFF0,FFFF,<PCI 16.0>,0003,0000`. This is particularly strange
+    // in that the new options were reordered around some other PCI device. It's
+    // not the boot order we set!
+    //
+    // So, to at least confirm we *can* modify the boot order in a stable way,
+    // make a somewhat less ambitious change: insert the duplicate boot options
+    // in the order directly after the option they are duplicates of. This seems
+    // to not get reordered.
+    let boot_option_bytes = read_efivar(&vm, &bootvar(boot_num)).await?;
+
+    // Finally, seeing a read-write `efivarfs` is not sufficient to know that
+    // writes to EFI variables will actually stick. For example, an Alpine live
+    // image backed by an ISO 9660 filesystem may have an EFI System Partition
+    // and `efivarfs`, but certainly cannot persist state and will drop writes
+    // to EFI variables.
+    //
+    // Check for this condition and exit early if the guest OS configuration
+    // will not let us perform a useful test.
+    write_efivar(&vm, &bootvar(0xfff0), &boot_option_bytes).await?;
+    let reread = read_efivar(&vm, &bootvar(0xfff0)).await?;
+    if reread.is_empty() {
+        warn!(
+            "guest environment drops EFI variable writes! \
+            exiting test WITHOUT VALIDATING ANYTHING"
+        );
+        return Ok(());
```

Comment:
I might `phd_skip!` in this case so that the "skipped" bucket gets incremented instead of the "passed" bucket in the runner output.

Reply:
nice! yes, that's much better.